tflite2onnx/op/reshape.py

```diff
@@ -72,17 +73,135 @@ def parse(self):
+        # holds a special data layout semantic,
+        # we need to insert Transpose before (or after) the Reshape operator
+        # https://github.com/jackwish/tflite2onnx/issues/28
+        # An example for inserting `Transpose` before `Reshape`
+        #    ------
+        #   | Conv |
+        #    ------
+        #      |
+        #      |  to_transpose (Original input of `Reshape`)
+        #      |  (e.g. NCHW)
+        #   -------
+        #  | Trans |  e.g. perm: (0, 2, 3, 1)
+        #   -------
+        #      |
+        #      |  transposed (New created tensor)
+        #      |  (e.g. NHWC)
+        #   --------
+        #  | Reshape |
+        #   --------
+        #
+
+        assert(self.status.parsed)
+        to_transpose = self.inputs[0]
+
+        transposed_name = 'TFLITE2ONNX_Transposed_%s' % to_transpose.name
+        transposed = self.TFactory.getWithRef(to_transpose, transposed_name, True)
+
+        # Construct the layout as original input of `Reshape`
+        # e.g. source: NHWC, target: NCHW
+        # Exchange the source and target layout for layout preservation
+        layout = Layout(to_transpose.layout.target, to_transpose.layout.source)
+        transposed.shape = layout.transform(to_transpose.shape)
+        transposed.setParsed()
+
+        # Construct the additional transpose before `Reshape`
+        trans = Transpose(self.TFactory, -1)
+        # e.g. From NCHW to NHWC
+        trans.attrs['perm'] = layout.perm
+
+        trans.inputs.append(to_transpose)
+        to_transpose.replaceConsumer(self, trans)
+        self.replaceInput(to_transpose, transposed)
+
+        trans.outputs.append(transposed)
+        transposed.addProducer(trans)
+        transposed.addConsumer(self)
+        trans.setParsed()
+
+        self.pre.append(trans)
+
+    def preserveOutputSpatialSemantic(self):
+        # If the input (or output) tensor of Reshape
+        # holds a special data layout semantic,
+        # we need to insert Transpose before (or after) the Reshape operator
+        # https://github.com/jackwish/tflite2onnx/issues/28
+
+        # An example for inserting `Transpose` after `Reshape`
+        #   -------
+        # | Reshape |
+        #   -------
+        #      |
+        #      |  to_transpose (New created tensor)
+        #      |  (e.g. NHWC)
+        #   -------
+        #  | Trans |  e.g. perm: (0, 3, 1, 2)
+        #   -------
+        #      |
+        #      |  transposed (Original `Reshape` output)
+        #      |  (e.g. NCHW)
+        #    ------
+        #   | Conv |
+        #    ------
+        #
+
+        assert(self.status.parsed)
+        transposed = self.outputs[0]
+
+        transposed_new_name = 'TFLITE2ONNX_ToTransposed_%s' % transposed.name
+        to_transpose_name = transposed.name
+        # Rename the transposed tensor
+        # because it will be the output of the new `transpose`
+        self.TFactory.registery[transposed_new_name] = \
```

Comment:
I mean for `[Reshape] -- <T_nhwc> -- [Conv]` in TFLite, after _layout propagation_ we got `[Reshape] -- <T_nchw> -- [Conv]` where we need to change it into `[Reshape] -- <T_nhwc> -- [Transpose] -- <T_nchw> -- [Conv]`. So we are building `<T_nhwc> -- [Transpose]` and insert it into the graph. We can simply name the `<T_nhwc>` as `ToTranspose`. The _to_ means this tensor is going to transpose.

Basically, the tensors and operators that start with a `TFLITE2ONNX` prefix are ones that we added to the graph as the helper, while others are converted from the TFLite model.

So maybe we can switch the name here in which way we don't need to modify the name of the existed tensors.

Reply:
Ah, I gotchu. The motivation to use the hacky way here is to avoid the issue that two nodes have the same name. 
The operator is named after its first output tensor in the `setParsed` ( before the  _layout propagation_ I think). So the `Reshape` has its name after `<T_nchw>`. Then the newly inserted 'Transpose' will also be named after `<T_nchw>` when `setParsed`, which makes the `Reshape` and `Transpose` the same name. So either the output tensor of `Transpose` or the operator `Reshape` should be renamed.

According to the convention you have mentioned:

> Basically, the tensors and operators that start with a TFLITE2ONNX prefix are ones that we added to the graph as the helper, while others are converted from the TFLite model.

I think it might be better to rename `Transpose` directly (although still a bit hacky lol ) because it is also a helper operator. How do you think about it?

Reply:
That's a good point, I don't even remember that we have used the name of the first tensor as node name. We didn't name the node at the beginning because TFLite operators are not named. But named operators do improve debug experience so we just named it with that policy which may require double consideration.

Renaming `Transpose` sounds good for me :)